engine/engine.3
.\"	$Csoft: agar.3,v 1.40 2004/02/26 10:34:58 vedge Exp $
.\"
.\" Copyright (c) 2001, 2002, 2003 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$OpenBSD: mdoc.template,v 1.6 2001/02/03 08:22:44 niklas Exp $
.\"
.Dd NOVEMBER 26, 2001
.Dt AGAR 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm engine
.Nd agar engine initialization and globals
.Sh SYNOPSIS
.Bd -literal
#include <engine/engine.h>
.Ed
.Sh DESCRIPTION
.Pp
These functions are responsible for the initialization of Agar's internals,
the initialization of SDL (and possibly OpenGL), the selection of the video
mode, the font engine and the attachment of input devices.
.Pp
.nr nS 1
.Ft "int"
.Fn engine_preinit "const char *progname"
.Pp
.Ft "void"
.Fn engine_set_gfxmode "enum gfx_engine mode"
.Pp
.Ft "void"
.Fn engine_set_textmode "enum text_engine mode"
.Pp
.Ft "int"
.Fn engine_init "void"
.nr nS 0
.Pp
The
.Fn engine_preinit
function initializes Agar's internal data structures as well as SDL.
Agar applications should invoke
.Fn engine_preinit
first, with the name of the application as the
.Fa progname
argument.
By default, Agar will use
.Pa ~/.<progname>
for the application's data files.
.Pp
The
.Fn engine_init
function sets the video mode, initializes the font engine and input devices.
.Pp
Between the calls to
.Fn engine_init
and
.Fn engine_preinit ,
several parameters may be set, such as the font engine, the font directory,
the default font and the graphic mode to use.
.Pp
The
.Fn engine_set_gfxmode
selects the graphic mode.
The graphic mode is mostly a hint to the Agar GUI which tells it how to
optimize certain GUI operations depending on your expected use of the
frame buffer outside of the GUI (ie. in the background).
This setting defines, for instance, the way the background is redrawn after
a window displacement or resizing.
Available graphic modes include:
.Bd -literal
enum gfx_engine {
	GFX_ENGINE_GUI,		/* Direct video/OpenGL, solid background */
	GFX_ENGINE_TILEBASED	/* Direct video/OpenGL, map background */
};
.Ed
.Pp
The
.Dv GFX_ENGINE_GUI
mode asserts that the background is a solid color.
When windows are moved or resized under this mode, Agar uses
.Xr SDL_FillRect 3
(or
.Xr glRecti 3
in OpenGL mode) to redraw parts of the background that need updating.
.Pp
The
.Dv GFX_ENGINE_TILEBASED
mode assumes that the background is the rendering of a
.Xr map 3 .
Therefore, it uses the
.Fn draw
operation of the map being displayed for redrawing the background.
A map must be always visible under this mode, even if it is empty.
Setting the background map to NULL causes the application to exit.
.Pp
The
.Dv GFX_ENGINE_TILEBASED
mode is preferable for applications such as 2D games, since it also optimizes
the way animations are processed.
.Pp
The
.Fn engine_set_textmode
function selects the font engine among:
.Bd -literal
enum text_engine {
	TEXT_ENGINE_TTF,		/* FreeType */
	TEXT_ENGINE_BITMAP,		/* Fixed-size bitmap */
	TEXT_ENGINE_PIXMAP		/* Fixed-size pixmap */
};
.Ed
.Pp
At the time, only
.Dv TEXT_ENGINE_TTF
is implemented.
.Dv TEXT_ENGINE_TTF
obtains its glyphs from the FreeType library.
.Sh INCLUDE FILES
The
.Pa Aq compat/*.h
and
.Pa Aq config/*.h
includes should come first, followed by
.Pa Aq engine/engine.h ,
which itself includes:
.Pp
.Bl -tag -width "<engine/loader/integral.h> " -compact
.It Pa Aq pthread.h
POSIX threads, if available and enabled.
.It Pa Aq SDL.h
Software DirectMedia Layer.
.It Pa Aq GL/GL.h
OpenGL interface, if available and enabled.
.It Pa Aq compat/queue.h
4.4BSD
.Xr queue 3
definitions.
.It Pa Aq compat/strlcpy.h
For
.Xr strlcpy 3 .
.It Pa Aq compat/strlcat.h
For
.Xr strlcat 3 .
.It Pa Aq compat/snprintf.h
For
.Xr snprintf 3 .
.It Pa Aq compat/vsnprintf.h
For
.Xr vsnprintf 3 .
.It Pa Aq compat/vasprintf.h
For
.Xr vasprintf 3 .
.It Pa Aq compat/asprintf.h
For
.Xr asprintf 3 .
.It Pa Aq compat/strsep.h
For
.Xr strsep 3 .
.It Pa Aq engine/loader/netbuf.h
For the
.Ft struct netbuf
definition, required by load/save function prototypes.
.It Pa Aq engine/loader/integral.h
For saving integers.
.It Pa Aq engine/loader/real.h
For saving real numbers.
.It Pa Aq engine/loader/string.h
For saving text.
.It Pa Aq engine/loader/version.h
For versioning.
.It Pa Aq engine/error/error.h
Generic error handling.
.It Pa Aq engine/object.h
Generic
.Xr object 3
types and operations.
.It Pa Aq engine/event.h
For the
.Ft union evarg
definition and generic
.Xr event 3
handling.
.El
.Sh SEE ALSO
.Xr object 3 ,
.Xr input 3 ,
.Xr map 3 ,
.Xr view 3 ,
.Xr vg 3 ,
.Xr widget 3 ,
.Xr window 3
.Pp
.Bd -literal
http://www.libsdl.org/
http://www.opengl.org/
.Ed


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

trunk/engine/engine.c
/*	$Csoft: engine.c,v 1.166 2005/10/02 09:41:08 vedge Exp $	*/

/*
 * Copyright (c) 2001, 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config/have_x11.h>
#include <config/have_setlocale.h>
#include <config/localedir.h>
#include <config/version.h>
#include <config/network.h>

#include <compat/setenv.h>
#include <compat/strlcat.h>

#include <engine/engine.h>
#include <engine/config.h>
#include <engine/view.h>
#include <engine/typesw.h>

#ifdef MAP
#include <engine/map/map.h>
#endif

#ifdef EDITION
#include <engine/map/mapedit.h>
#endif

#ifdef NETWORK
#include <engine/rcs.h>
#endif

#include <engine/widget/widget.h>
#include <engine/widget/window.h>
#include <engine/widget/textbox.h>
#include <engine/widget/keycodes.h>
#include <engine/widget/primitive.h>
#include <engine/widget/cursors.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#ifdef HAVE_SETLOCALE
#include <locale.h>
#endif
#ifdef HAVE_X11
#include <SDL_syswm.h>
#endif

#ifdef THREADS
pthread_mutexattr_t agRecursiveMutexAttr;	/* Recursive mutex attributes */
#endif

const char *agProgName = "";
AG_Config *agConfig;
AG_Object *agWorld;
AG_Object agIconMgr;
void (*agAtexitFunc)(void) = NULL;
pthread_mutex_t agLinkageLock;
pthread_mutex_t agTimingLock;
int agServerMode = 0;

int
AG_InitCore(const char *progname, u_int flags)
{
	agProgName = progname;

#ifdef HAVE_SETLOCALE
	setlocale(LC_ALL, "");
#endif
#ifdef ENABLE_NLS
	bindtextdomain("agar", LOCALEDIR);
	textdomain("agar");
#endif

	AG_InitError();

#ifdef THREADS
	pthread_mutexattr_init(&agRecursiveMutexAttr);
	pthread_mutexattr_settype(&agRecursiveMutexAttr,
	    PTHREAD_MUTEX_RECURSIVE);
	pthread_mutex_init(&agLinkageLock, &agRecursiveMutexAttr);
	pthread_mutex_init(&agTimingLock, &agRecursiveMutexAttr);
#endif
	if (SDL_Init(SDL_INIT_TIMER|SDL_INIT_NOPARACHUTE) != 0) {
		AG_SetError("SDL_Init: %s", SDL_GetError());
		return (-1);
	}

	AG_InitTypeSw();
	
	agConfig = Malloc(sizeof(AG_Config), M_OBJECT);
	AG_ConfigInit(agConfig);
	AG_ObjectLoad(agConfig);

	printf("\n");
	printf(_("Architecture extensions: "));
	if (SDL_HasRDTSC()) { printf("RDTSC "); }
	if (SDL_HasMMX()) { printf("MMX "); }
	if (SDL_HasMMXExt()) { printf("MMXExt "); }
	if (SDL_Has3DNow()) { printf("3DNow "); }
	if (SDL_Has3DNowExt()) { printf("3DNowExt "); }
	if (SDL_HasSSE()) { printf("SSE "); }
	if (SDL_HasSSE2()) { printf("SSE2 "); }
	if (SDL_HasAltiVec()) { printf("AltiVec "); }
	printf(".\n");

	agWorld = AG_ObjectNew(NULL, "world");
	AG_ObjectRemain(agWorld, AG_OBJECT_REMAIN_DATA);
	return (0);
}

int
AG_InitVideo(int w, int h, int bpp, u_int flags)
{
	extern int agBgPopupMenu;

#ifdef HAVE_X11
	setenv("SDL_VIDEO_X11_WMCLASS", agProgName, 1);
#endif
	if (SDL_InitSubSystem(SDL_INIT_VIDEO) != 0) {
		AG_SetError("SDL_INIT_VIDEO: %s", SDL_GetError());
		return (-1);
	}
	SDL_WM_SetCaption(agProgName, agProgName);

	if ((agVideoInfo = SDL_GetVideoInfo()) != NULL) {
		char accel[2048];
		char unaccel[2048];
		size_t size = 2048;

		agVideoFmt = agVideoInfo->vfmt;

		if (agVideoInfo->wm_available) {
			printf(_("Window manager is available.\n"));
		} else {
			printf(_("Window manager is unavailable.\n"));
		}
			
		if (agVideoInfo->hw_available) {
			printf(_("Hardware surfaces are available.\n"));
		} else {
			printf(_("Hardware surfaces are unavailable.\n"));
		}
		
		accel[0] = '\0';
		unaccel[0] = '\0';

		strlcat(agVideoInfo->blit_hw ? accel : unaccel,
		    _("\tSDL hardware blits\n"), size);
		strlcat(agVideoInfo->blit_hw_CC ? accel : unaccel,
		    _("\tSDL hardware->hardware colorkey blits\n"), size);
		strlcat(agVideoInfo->blit_hw_A ? accel : unaccel,
		    _("\tSDL hardware->hardware alpha blits\n"), size);
		strlcat(agVideoInfo->blit_sw ? accel : unaccel,
		    _("\tSDL software->hardware blits\n"), size);
		strlcat(agVideoInfo->blit_sw_CC ? accel : unaccel,
		    _("\tSDL software->hardware colorkey blits\n"), size);
		strlcat(agVideoInfo->blit_sw_A ? accel : unaccel,
		    _("\tSDL software->hardware alpha blits\n"), size);
		strlcat(agVideoInfo->blit_fill ? accel : unaccel,
		    _("\tSDL color fills\n"), size);

		if (accel[0] != '\0')
			printf(_("Accelerated operations:\n%s"), accel);
		if (unaccel[0] != '\0')
			printf(_("Unaccelerated operations:\n%s"), unaccel);

		printf("\n");
	}

	if (flags & (AG_VIDEO_OPENGL|AG_VIDEO_OPENGL_OR_SDL)) {
#ifdef HAVE_OPENGL
		AG_SetBool(cfg, "view.opengl", 1);
#else
		if ((flags & AG_VIDEO_OPENGL_OR_SDL) == 0)
			fatal("Agar OpenGL support is not compiled in");
#endif
	}
	if (AG_ViewInit(w, h, bpp, flags) == -1 || AG_TextInit() == -1) {
		return (-1);
	}
	AG_ColorsInit();
	AG_InitPrimitives();
	AG_CursorsInit();

	if (flags & AG_VIDEO_BGPOPUPMENU) { agBgPopupMenu = 1; }
	
	AG_ObjectInit(&agIconMgr, "object", "IconMgr", NULL);
	AG_WireGfx(&agIconMgr, "/engine/icons/icons");
	return (0);
}

int
AG_InitConfigWin(u_int flags)
{
	AG_ConfigWindow(agConfig, flags);
	return (0);
}

/* Initialize graphics and input devices. */
int
AG_InitInput(u_int flags)
{
	extern int agKbdUnicode;			/* config.c */
	int i, n, njoys;

	SDL_EnableUNICODE(agKbdUnicode || (flags & AG_FORCE_UNICODE));

	if (AG_Bool(agConfig, "input.joysticks") ||
	    (flags & AG_FORCE_JOYSTICK)) {
		if (SDL_InitSubSystem(SDL_INIT_JOYSTICK) == 0) {
			n = SDL_NumJoysticks();
			for (i = 0, njoys = 0; i < n; i++) {
				if (SDL_JoystickOpen(i) != NULL)
					njoys++;
			}
			if (njoys > 0)
				SDL_JoystickEventState(SDL_ENABLE);
		}
	}
	return (0);
}

Uint8
AG_MouseGetState(int *x, int *y)
{
	Uint8 rv;

	rv = SDL_GetMouseState(x, y);
#if defined(__APPLE__) && defined(HAVE_OPENGL)
	if (agView->opengl && y != NULL)
		*y = agView->h - *y;
#endif
	return (rv);
}

int
AG_InitNetwork(u_int flags)
{
#if defined(DEBUG) && defined(NETWORK) && defined(THREADS)
	if ((flags & AG_INIT_DEBUG_SERVER) && agServerMode) {
		extern int AG_DebugServerStart(void);
		AG_DebugServerStart();
	}
#endif
#if defined(NETWORK)
	if (flags & AG_INIT_RCS)
		AG_RcsInit();
#endif
	return (0);
}

/* Register a function to invoke in AG_Quit(). */
void
AG_AtExitFunc(void (*func)(void))
{
	agAtexitFunc = func;
}

/* Request a graceful shutdown of the application. */
void
AG_Quit(void)
{
	SDL_Event nev;

	nev.type = SDL_QUIT;
	SDL_PushEvent(&nev);
}

/* Immediately clean up and exit the application. */
void
AG_Destroy(void)
{
	if (agAtexitFunc != NULL)
		agAtexitFunc();

#if defined(MAP) && defined(EDITION)
	if (agEditMode)
		AG_ObjectSave(&agMapEditor);
#endif
#ifdef NETWORK
	AG_RcsDestroy();
#endif

	AG_ObjectDestroy(agWorld);
	AG_TextDestroy();
	AG_ViewDestroy();
	
	AG_ColorsDestroy();
	AG_CursorsDestroy();

	AG_ObjectDestroy(agConfig);
	Free(agConfig, M_OBJECT);

/*	pthread_mutex_destroy(&agLinkageLock); */
	pthread_mutex_destroy(&agTimingLock);

	AG_DestroyError();
	AG_DestroyTypeSw();
	SDL_Quit();
	exit(0);
}

